pkg/prometheus/resource_selector.go

```diff
@@ -1033,8 +1033,25 @@ func (rs *ResourceSelector) validateHTTPSDConfigs(ctx context.Context, sc *monit
 }
 
 func (rs *ResourceSelector) validateDNSSDConfigs(sc *monitoringv1alpha1.ScrapeConfig) error {
+	promVersion := operator.StringValOrDefault(rs.p.GetCommonPrometheusFields().Version, operator.DefaultPrometheusVersion)
+	version, err := semver.ParseTolerant(promVersion)
+	if err != nil {
+		return fmt.Errorf("failed to parse Prometheus version: %w", err)
+	}
 	for i, config := range sc.Spec.DNSSDConfigs {
 		if config.Type != nil {
+			if *config.Type == "NS" && version.LT(semver.MustParse("2.49.0")) {
+				return fmt.Errorf("type 'NS' is only supported for Prometheus version >= 2.49.0")
+			}
+
+			if *config.Type == "A" || *config.Type == "AAAA" && version.LT(semver.MustParse("0.16.0")) {
+				return fmt.Errorf("type 'A' or 'AAAA' is only supported for Prometheus version >= 0.16.0")
+			}
+
+			if *config.Type == "MX" && version.LT(semver.MustParse("2.38.0")) {
+				return fmt.Errorf("type 'MX' is only supported for Prometheus version >= 2.38.0")
+			}
+
```

Comment:
What if we create a map here?

```go
compatibilityMatrix := map[config.Type]semver.Version{
  "NS": semver.MustParse("2.49.0")
  // complete with all others
}

for i, config := range sc.Spec.DNSSDConfigs {
  if v, ok := compatibilityMatrix[config.Type]; ok {
    if version.LT(v) {
      return fmt.error ....
    }
  }
}
```

code written from mobile, so not really thought through, but maybe it's more readable than a ton of ifs? WDYT?

Reply:
I was definitely unsure of the number of if's  there, was hoping someone might suggest something better. I will check this out.